internal/csi-addons/rbd/volumegroup.go

```diff
@@ -421,10 +426,55 @@ func (vs *VolumeGroupServer) ModifyVolumeGroupMembership(
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 
-	// Skip modification of group if it's secondary
-	if !vgrMirrorInfo.IsPrimary() {
-		log.DebugLog(ctx, "skipping modification of group, as it is in secondary state")
-		return &volumegroup.ModifyVolumeGroupMembershipResponse{}, nil
+	// If the mirroring is not enabled and we are modifying the group, that means
+	// we are doing a cleanup and we shouldn't be checking for mirror status
+	isMirroringEnabled := false
+	if vgrMirrorInfo.GetState() == librbd.MirrorImageEnabled.String() {
```

Comment:
This part of the function is really quite complex, it can benefit a lot from splitting this out in one or more helper functions. Specially the `librbd` checking should be done somewhere in the `internal/rbd` package, the `internal/csi-addons/rbd` package should only use go-ceph indirectly. However, it seems that `internal/csi-addons/rbd/replication.go` violates this rule already too :disappointed: You can keep this for now, but it does need improving before it can be accepted in upstream.

Reply:
Once, we have dynamic grouping available in ceph. We can remove all these checks, the checks are in place only to support dynamic grouping and locking issues in ceph rbd.